add headers impl

Ethan-Arrowood opened the pull request:
This PR adds a Headers class implementation based on the WHATWG Fetch Spec. It leaves off many pieces of the spec due to the non-browser environment. Additionally, it deviates from the spec slightly for sake of performance improvements.

While it maintains support for spec based initialization, folks looking for most-performant solution should utilize a flattened array.

Header entries are kept sorted at all times so all read-based use cases are fast as possible. Writing is also kept fast by using a probe-based binary search/insert.

This code is based on my work in [undici-fetch](https://github.com/ethan-Arrowood/undici-fetch). The headers api is the most stable of the pieces and has no dependencies on other things like whatwg readablestreams, so it makes for a good-first-pr for fetch in Node! I also believe this class could be utilized by other HTTP aspects of the Node API - but the initial goal is relation to Fetch.

jasnell commented:
Wooo! Obviously still needs the docs. When those are added, let's be sure to mark this experimental for now.

devsnek commented:
Any reason this isn't just a map of arrays?

```js
function validateName(name) {
  if (!name.match(/^[!#$%&'*+\-.^`|~\w]+$/)) {
    throw new ERR_OUT_OF_RANGE('name', 'a valid header name', name);
  }
  return toUSVString(name).toLowerCase();
}

function validateValue(value) {
  value = toUSVString(value)
    .replace(/^[\r\n\t ]+/g, '')
    .replace(/[\r\n\t ]+$/, '');
  if (/[\0\r\n]/.test(value)) {
    throw new ERR_OUT_OF_RANGE('name', 'a valid header value', value);
  }
  return value;
}

class Headers {
  #headers = new SafeMap();

  constructor(init = undefined) {
    if (init) {
      if (init instanceof Headers) {
        this.#headers = new SafeMap([...init.#headers]);
      } else if (ArrayIsArray(init)) {
        for (const header of init) {
          if (header.length !== 2) {
            throw new ERR_OUT_OF_RANGE();
          }
          this.append(header[0], header[1]);
        }
      } else {
        for (const pair of ObjectEntries(init)) {
          this.append(pair[0], pair[1]);
        }
      }
    }
  }

  append(name, value) {
    value = validateValue(value);
    name = validateName(name);
    const existing = this.#headers.get(name.toLowerCase());
    if (existing) {
      existing.push(value);
    } else {
      this.#headers.set(name.toLowerCase(), [value]);
    }
  }

  delete(name) {
    name = validateName(name);
    this.#headers.delete(name);
  }

  get(name) {
    name = validateName(name);
    const values = this.#headers.get(name);
    if (!values) {
      return null;
    }
    return values.join(', ');
  }

  has(name) {
    name = validateName(name);
    return this.#headers.has(name);
  }

  set(name, value) {
    value = validateValue(value);
    name = validateName(name);
    this.#headers.set(name, [value]);
  }
}
```

Ethan-Arrowood commented:
> Any reason this isn't just a map of arrays?

I believe this was previously considered, but a flat array of strings was assumed to be more efficient in regards to space and speed.

targos commented:
How do we want to expose this and the other fetch features ( before they become globals)?

devsnek commented:
> would require some benchmarking and more research on arrays vs Map

benchmarks would be good.

timdp commented:
> > Any reason this isn't just a map of arrays?
> 
> I believe this was previously considered, but a flat array of strings was assumed to be more efficient in regards to space and speed.
> 
> Emphasis on _assumed_; would require some benchmarking and more research on arrays vs Map

Yeah, I remember assuming that. 🙂 It'd indeed be interesting to benchmark!

bl-ue commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,212 @@
+/**
+ * This algorithm is based off of
+ * https://www.tbray.org/ongoing/When/200x/2003/03/22/Binary
+ * It only operates on the even indexes of the array (the header names) by only
```
```suggestion
 * It only operates on the even indices of the array (the header names) by only
```

Trott replied:
We follow the Microsoft Style Guide and [so `indexes` is the way to go here](https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/i/index-indexes-indices).

RaisinTen commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,212 @@
+ */
+function binarySearch(arr, val) {
+  let low = 0;
+  let high = arr.length / 2;
```
```suggestion
  let high = MathFloor(arr.length / 2);
```
The source is in Java, so this does integer division. That's not the case in JS, so explicitly passing this through a `floor()`.

RaisinTen commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,212 @@
+} = require('internal/errors');
+
+const {
+  ObjectEntries,
+  ArrayIsArray,
+  StringPrototypeToLocaleLowerCase,
+  StringPrototypeReplace,
+  Symbol,
+  SymbolIterator,
+  Array,
```
```suggestion
  Array,
  ArrayIsArray,
  MathFloor,
  ObjectEntries,
  StringPrototypeReplace,
  StringPrototypeToLocaleLowerCase,
  Symbol,
  SymbolIterator,
```
lexicographical order + `MathFloor`

Ethan-Arrowood commented:
@devsnek here is some preliminary benchmarks: https://gist.github.com/Ethan-Arrowood/e132fb72e5ef4156dfc0de3b80ef3c16

A `Map` based implementation is only slightly faster in some cases and in some scenarios. It seem that the overhead of the `Map` api is not faster than just using arrays; and that the binarySort is actually quite fast!

The iteration of `Map` impl could be faster if the `entries` method was implemented differently; such as sorting as the `headers` array is created from the Map iterator.

Ethan-Arrowood commented:
Alright this PR is ready for review. All docs and tests from undici-fetch have been copied over (and transformed to the Node way). This is my first large API addition to Node so I'm probably missing some things 🤷‍♂️

aduh95 commented on doc/api/fetch.md:
```diff
@@ -0,0 +1,225 @@
+# Fetch
+
```
```suggestion

<!--introduced_in=REPLACEME-->

> Stability: 1 - Experimental
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,212 @@
+  while (high > low) {
+    const mid = (high + low) >>> 1;
+
+    if (val.localeCompare(arr[mid * 2]) > 0) {
```
```suggestion
    if (StringPrototypeLocaleCompare(val, arr[mid * 2]) > 0) {
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,212 @@
+
+function normalizeAndValidateHeaderValue(name, value) {
+  // https://fetch.spec.whatwg.org/#concept-header-value-normalize
+  const normalizedHeaderValue = StringPrototypeReplace(
+    value,
+    /^[\n\t\r\x20]+|[\n\t\r\x20]+$/g,
+    ''
+  );
```
```suggestion
  const normalizedHeaderValue = RegExpPrototypeSymbolReplace(
    /^[\n\t\r\x20]+|[\n\t\r\x20]+$/g,
    value,
    ''
  );
```

Ethan-Arrowood replied:
Any reason for this recommendation? Is RegExp replace faster than String replace?

aduh95 replied:
`%String.prototype.replace%` simply calls `regex[Symbol.replace](this, replacer)`, so you can skip a step by using the RegExp method directly.

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,212 @@
+    }
+  }
+
+  *[SymbolIterator]() {
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      yield [this[kHeadersList][index], this[kHeadersList][index + 1]];
+    }
+  }
+}
```
I think we should mimic how [`Map.prototype[@@iterator]`](https://tc39.es/ecma262/multipage/keyed-collections.html#sec-map.prototype-@@iterator) is defined:
```suggestion
}

Headers.prototype[SymbolIterator] = Headers.prototype.entries;
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  *entries() {
+    yield* this;
```
```suggestion
    for (let index = 0; index < this[kHeadersList].length; index += 2) {
      yield [this[kHeadersList][index], this[kHeadersList][index + 1]];
    }
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,212 @@
+
+module.exports = {
+  Headers,
+  kHeadersList,
+  binarySearch
```
nit: ASCII order
```suggestion
  binarySearch,
  kHeadersList,
```

aduh95 reviewed (commented):
You need to run the linter on the markdown file, there are a few CI errors that are reported

aduh95 commented on test/parallel/test-headers.js:
```diff
@@ -0,0 +1,376 @@
+
+{
+  // Init fails silently when initialized with BoxedPrimitives
+  try {
+    new Headers(new Number());
+    new Headers(new Boolean());
+    new Headers(new String());
+  } catch (error) {
+    common.mustNotCall(error);
+  }
```
```suggestion
  [new Number(), new Boolean(), new String()].forEach((arg) => new Headers(arg));
```

aduh95 commented on test/parallel/test-headers.js:
```diff
@@ -0,0 +1,376 @@
+
+{
+  // Init fails silently if function or primitive is passed
+  try {
+    new Headers(Function);
+    new Headers(function() {});
+    new Headers(1);
+    new Headers('test');
+  } catch (error) {
+    common.mustNotCall(error);
+  }
```
```suggestion
  [
    Function,
    function() {},
    1,
    'test',
  ].forEach((arg) => new Headers(arg));
```

aduh95 commented on test/parallel/test-headers.js:
```diff
@@ -0,0 +1,376 @@
+    'test-name-3',
+    'test-value-3'
+  ]);
+  assert.doesNotThrow(() => headers.delete('test-name-2'))
```
```suggestion
  headers.delete('test-name-2');
```

aduh95 commented on test/parallel/test-headers.js:
```diff
@@ -0,0 +1,376 @@
+  tests.forEach(({ input: [list, target], expected, message }) => {
+    assert.deepStrictEqual(expected, binarySearch(list, target), message)
+  })
+}
```
```suggestion
}

```

bl-ue reviewed (commented):
Which is prefered for indention in the docs: tabs or spaces, and how many?  I noticed you're using both a single tab and two spaces for indentation here.

Ethan-Arrowood commented:
How do I get the node linter to auto fix?

aduh95 commented:
> How do I get the node linter to auto fix?

You can do `make lint-js-fix`.

Ethan-Arrowood commented:
And how do I deal with strings that exceed 80 characters in length? The linter isn't fixing those automatically

targos commented:
Unfortunately, ESLint doesn't have an auto-fix for this rule. I usually reformat the code with Prettier in these cases.

Ethan-Arrowood commented:
> I usually reformat the code with Prettier in these cases.

Do you just `npx prettier <file_name>`?

aduh95 commented on doc/api/fetch.md:
```diff
@@ -0,0 +1,228 @@
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+## Class: `fetch.Headers`
```
```suggestion
> Stability: 1 - Experimental

## Class: `fetch.Headers`
```

aduh95 commented on doc/api/fetch.md:
```diff
@@ -0,0 +1,228 @@
+> Stability: 1 - Experimental
+## Class: `fetch.Headers`
+
+Represents a WHATWG Fetch Spec [Headers Class](https://fetch.spec.whatwg.org/#headers-class)
```
```suggestion
Represents a WHATWG Fetch Spec [Headers Class](https://fetch.spec.whatwg.org/#headers-class).
```

aduh95 reviewed (commented):
As a user, how do I have access to this class? Unless I'm missing something, it's internal-only with the current implementation. If we are creating a new doc page, the end-users should be able to access it somehow without using `--expose-internals`: either as a global (in which case, it would make this `semver-major`), either as a new core module.

aduh95 commented on doc/api/fetch.md:
```diff
@@ -0,0 +1,241 @@
+### `new Headers([init])`
+
+* `init` {Headers | Iterable<[string, string]> | string[] | Record<string,
+  string>} Initial header list to be cloned into the new instance
```
```suggestion
  string>} Initial header list to be cloned into the new instance.
```

aduh95 commented on doc/api/fetch.md:
```diff
@@ -0,0 +1,241 @@
+
+Retrieves a header entry. If the entry _name_ has multiple values, they are
+returned as a string joined by `','` characters. If the _name_ does not exist,
+this method returns null.
```
```suggestion
this method returns `null`.
```

Ethan-Arrowood commented:
> Unless I'm missing something, it's internal-only with the current implementation.

Nope; just my own lack-of-understanding. How do i expose this as an API in Node? I couldn't find anything in the contribution docs for it

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,213 @@
+    }
+  } else if (!isBoxedPrimitive(object)) {
+    // Object of key/value entries
+    for (const { 0: name, 1: value } of ObjectEntries(object)) {
+      headers.append(name, value);
+    }
```
You can avoid unsafe Array iteration here.
```suggestion
    const entries = ObjectEntries(object);
    for (let i = 0; i < entries.length; i++) {
      headers.append(entries[i][0], entries[i][1]);
    }
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,213 @@
+  }
+
+  * keys() {
+    for (const header of this) {
+      yield header[0];
+    }
+  }
+
+  * values() {
+    for (const header of this) {
+      yield header[1];
+    }
```
Wouldn't it be more efficient to avoid the array creation?
```suggestion
    for (let index = 0; index < this[kHeadersList].length; index += 2) {
      yield this[kHeadersList][index];
    }
  }

  * values() {
    for (let index = 1; index < this[kHeadersList].length; index += 2) {
      yield this[kHeadersList][index];
    }
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,214 @@
+function fill(headers, object) {
+  if (kHeadersList in object) {
+    // Object is instance of Headers
+    headers[kHeadersList] = new Array(...object[kHeadersList]);
```
Is this simply cloning the array? Using `.slice` seems to be vastly more efficient for that: https://jsben.ch/M2mgy
```suggestion
    headers[kHeadersList] = ArrayPrototypeSlice(object[kHeadersList]);
```

aduh95 commented:
> > Unless I'm missing something, it's internal-only with the current implementation.
> 
> Nope; just my own lack-of-understanding. How do i expose this as an API in Node? I couldn't find anything in the contribution docs for it

I don't think this process ever been properly documented indeed. You can take inspiration from https://github.com/nodejs/node/pull/37970, let us know if you stumble upon a blocking point.

Ethan-Arrowood commented on test/parallel/test-headers.js:
```diff
@@ -0,0 +1,491 @@
+    function() {},
+    1,
+    'test',
+  ].forEach((arg) => new Headers(arg));
```
This test case is failing cause of this line:

```js
validateObject(init, 'init', { allowArray: true }) === undefined)
```

Any ideas how to fix this?

Ethan-Arrowood replied:
Do we have something like "isPrimitive"?

Ethan-Arrowood replied:
Resolving for a typeof check for now. welcome to other ideas

aduh95 replied:
A better resolution would be to not allow primitives, like browsers do IMO.

Ethan-Arrowood replied:
removed this so it follows spec

jasnell commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,225 @@
+}
+
+class Headers {
+  constructor(init) {
+    this[kHeadersList] = [];
+
+    if (init &&
+      validateObject(init, 'init', { allowArray: true }) === undefined) {
+      fill(this, init);
+    }
+  }
```
I would simplify this. Default the value of `init` to `{}` and just call `validateObject()` -- `validateObject()` does not return any value so checking the return with `=== undefined` is not necessary.

```suggestion
  constructor(init = {}) {
    validateObject(init, 'init', { allowArray: true });
    this[kHeadersList] = [];
    fill(this, init);
  }
```

jasnell commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,225 @@
+}
+
+function fill(headers, object) {
+  if (kHeadersList in object) {
```
I'd abstract out a `isHeaders()` brand check function here that can be reused.

jasnell commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,225 @@
+    }
+  }
+
+  append(name, value) {
```
These members need brand checks... e.g.

```js
append(name, value) {
  if (!isHeaders(this))
    throw new ERR_INVALID_THIS('Headers');
  // ...
```

jasnell commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,225 @@
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList][index + 1] += `, ${normalizedValue}`;
+    } else {
+      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue);
```
Use ArrayPrototypeSlice here?

jasnell commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,225 @@
+    }
+  }
+}
+
```
Add in here...

```js
ObjectDefineProperties(Headers.prototype, {
  append: { enumerable: true },
  delete: { enumerable: true },
  get: { enumerable: true },
  has: { enumerable: true },
  set: { enumerable: true },
  keys: { enumerable: true },
  values: { enumerable: true },
  entries: { enumerable: true },
  forEach: { enumerable: true },
});
```

jasnell commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,225 @@
+        this
+      );
+    }
+  }
```
Should implement a custom inspect function in here so that the kHeadersList internal property is not directly exposed.

Ethan-Arrowood replied:
Not sure what you mean here. Do you have an example of "custom inspect function" I can read through?

Ethan-Arrowood replied:
Do you mean something like:

```js
const { inspect } = require('internal/util/inspect');

class Headers {
  [inspect.custom] () {
    return this[kHeadersList];
  }
}
```

jasnell replied:
Take a look in `lib/internal/webstreams/readablestreams` for an example. Look for the `[kInspect]` function.

jasnell commented:
@Ethan-Arrowood ... would you mind if I opened a PR against your branch? I have some more detailed changes I'd like to make here to get it closer to spec compliance?

Ethan-Arrowood commented:
Thank you for the review @jasnell 

> What I'm thinking instead is to implement a lower level "header list" class that does what I need that can eventually sit under this class (replacing the kHeaderList's use of an array.

I originally built something like this but the moment you do anything more than an array you lose some of the major speed and memory performance benefits that makes this implementation unique (and potentially viable). 

I was rather thinking of extracting some of the methods I'm using in this class and making them a sort of 'Headers List Utility Set' that you can then utilize in your own Headers class implementation and still reap the same perf benefits of an underlying array. We sorta already have that since the binary search / sort is its own function, and so are the normalize/validate functions.

Ethan-Arrowood commented:
Summary of changes today:
- incorporated @jasnell review
- fixed linting errors
- worked on docs; some things are still failing in `make doc-only` that I'm not sure how to fix

I can tell we are getting very close with this 🙌

aduh95 commented on lib/fetch.js:
```diff
@@ -0,0 +1,15 @@
+
+emitExperimentalWarning('fetch/headers');
+
+const {
+  Headers
+} = require('internal/fetch/headers');
```
```suggestion
const {
  Headers,
} = require('internal/fetch/headers');
```
or
```suggestion
const { Headers } = require('internal/fetch/headers');
```

aduh95 commented on lib/fetch.js:
```diff
@@ -0,0 +1,15 @@
+  Headers
+} = require('internal/fetch/headers');
+
+module.exports = {
+  Headers
+};
```
```suggestion
module.exports = {
  Headers,
};
```
or
```suggestion
module.exports = { Headers };
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,291 @@
+    ERR_INVALID_ARG_VALUE,
+    ERR_INVALID_HTTP_TOKEN,
+    ERR_INVALID_THIS,
+    ERR_HTTP_INVALID_HEADER_VALUE
+  }
```
nit
```suggestion
    ERR_HTTP_INVALID_HEADER_VALUE,
  },
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,291 @@
+  ArrayPrototypeSlice,
+  ArrayPrototypeSplice,
+  MathFloor,
+  ObjectEntries,
+  RegExpPrototypeSymbolReplace,
+  StringPrototypeLocaleCompare,
+  StringPrototypeToLocaleLowerCase,
+  Symbol,
+  SymbolIterator,
+  ObjectDefineProperties
```
nit: ASCII order
```suggestion
  ObjectDefineProperties,
  ObjectEntries,
  RegExpPrototypeSymbolReplace,
  StringPrototypeLocaleCompare,
  StringPrototypeToLocaleLowerCase,
  Symbol,
  SymbolIterator,
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,291 @@
+        }
+        headers.append(object[i][0], object[i][1]);
+      }
+    } else if (typeof object[0] === 'string' || Buffer.isBuffer(object[0])) {
```
Browsers don't seem to support this:

```js
new Headers(['test', 'value'])
// Chromium: Uncaught TypeError: Failed to construct 'Headers': The provided value cannot be converted to a sequence.
// Firefox: Uncaught TypeError: Headers constructor: Element of sequence<sequence<ByteString>> branch of (sequence<sequence<ByteString>> or record<ByteString, ByteString>) can't be converted to a sequence.
// Safari: TypeError: Value is not a sequence
```
Am I missing something or should this be removed? Also why `Buffer`?

Ethan-Arrowood replied:
This was the one major deviation from spec. Supporting a 1D array of headers proves to be the most efficient solution

jasnell replied:
This might be something that is worth bringing back to the WHATWG spec folks to see if 1D arrays *could* be supported. /cc @annevk @domenic

annevk replied:
It would be best if someone filed an issue against whatwg/fetch to discuss that, ideally with motivation. It seems a bit too special to me, but it might fly.

domenic replied:
I think the idea is to match JS's Map type, so probably first you'd want to get Ecma 262's Map updated.

domenic replied:
I'd also suggest not adding non-spec constructors if you want this capability but instead some Node-specific utility method e.g. `http.headersFrom1DArray()`. The way you've done it now promotes non-isomorphic code.

Ethan-Arrowood replied:
I'll open something in the fetch repo. I'm unclear how this parallels to ECMA Map though.

domenic replied:
The 2D array syntax for the Headers constructor is derived from the 2D array syntax for the Map constructor.

Ethan-Arrowood replied:
going to resolve this comment and we can continue the discussion in the whatwg/fetch repo

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,291 @@
+class Headers {
+  constructor(init = {}) {
+    // Fail silently on primitives
+    const typeofInit = typeof init;
+    if (typeofInit === 'number' ||
+     typeofInit === 'string' ||
+     typeofInit === 'boolean' ||
+     typeofInit === 'function') return;
```
Why fail silently? Browsers seem to be pretty vocal about it:

```js
new Headers(1); // TypeError
new Headers(1n); // TypeError
new Headers(''); // TypeError
new Headers(true); // TypeError
new Headers(Symbol()); // TypeError

new Headers(() => {}); // no type error
```

Ethan-Arrowood replied:
Ahh I must've misunderstood the spec

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,291 @@
+    }
+
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      callback.call(
```
```suggestion
      FunctionPrototypeCall(
        callback,
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,291 @@
+    }
+  }
+
+  [inspect.custom] () {
```
```js
const {
  customInspectSymbol,
} = require('internal/util');
```
```suggestion
  [customInspectSymbol]() {
```

aduh95 commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,291 @@
+Headers.prototype[SymbolIterator] = Headers.prototype.entries;
+
+ObjectDefineProperties(Headers.prototype, {
+  append: { enumerable: true },
+  delete: { enumerable: true },
+  get: { enumerable: true },
+  has: { enumerable: true },
+  set: { enumerable: true },
+  keys: { enumerable: true },
+  values: { enumerable: true },
+  entries: { enumerable: true },
+  forEach: { enumerable: true },
```
```js
const {
  kEnumerableProperty,
} = require('internal/webstreams/util');
```
```suggestion
  append: kEnumerableProperty,
  delete: kEnumerableProperty,
  get: kEnumerableProperty,
  has: kEnumerableProperty,
  set: kEnumerableProperty,
  keys: kEnumerableProperty,
  values: kEnumerableProperty,
  entries: kEnumerableProperty,
  forEach: kEnumerableProperty,
```
It doesn't have to happen in this PR, but it may be worth moving this utility somewhere that is not directly related to webstreams – although `fetch` and Web Streams are not too far away.

aduh95 commented on test/parallel/test-headers.js:
```diff
@@ -0,0 +1,492 @@
+
+const {
+  Headers,
+  kHeadersList,
+  binarySearch
```
nit: ASCII order
```suggestion
  binarySearch,
  kHeadersList,
```

aduh95 commented on test/parallel/test-headers.js:
```diff
@@ -0,0 +1,492 @@
+{
+  // Init fails silently when initialized with BoxedPrimitives
+  [
+    new Number(),
+    new Boolean(),
+    new String(),
```
Note that `typeof new Number === 'object'`, same for the other primitives.

jasnell commented on doc/api/index.md:
```diff
@@ -29,6 +29,7 @@
 * [Domain](domain.md)
 * [Errors](errors.md)
 * [Events](events.md)
+* [Fetch](fetch.md)
```
I am a bit concerned that folks will misinterpret this as being a full fetch implementation...

Ethan-Arrowood replied:
What would be the best way to introduce this without implementing all of fetch?

I could make a new `feature/fetch` branch and we can merge all fetch work to that until its completely ready. I'm really eager to start integrating `undici-fetch` with the `streams/web` api

jasnell commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,291 @@
+Headers.prototype[SymbolIterator] = Headers.prototype.entries;
+
+ObjectDefineProperties(Headers.prototype, {
+  append: { enumerable: true },
```
Take a look in the `ReadableStream` implementation for how this is handled there. Basically, create a null-prototype object that can be reused here to prevent the possibility of a prototype pollution issue.

Ethan-Arrowood commented on lib/internal/fetch/headers.js:
```diff
@@ -0,0 +1,289 @@
+    fill(this, init);
+  }
+
+  append(name, value) {
```
I think we have to add an argument length check for the public methods. When you call `.append()` or `.append('foo')` on Chrome/Edge/FireFox/Safari you get a TypeError about invalid arg length / missing arguments. Should I use the `ERR_MISSING_ARGS` to accomplish this?

ronag commented:
tests fail

Ethan-Arrowood commented:
Do we want to close this PR in favor of adding Fetch via Undici instead? Fetch in Undici core is now under active development by @ronag

Ethan-Arrowood commented:
Closing in favor of undici fetch 😉
